### PR TITLE
Blog: rebase and reformat git branches

### PR DESCRIPTION
**Pull request by stagnation**

### CODE REVIEW

File: blog/2023-12-10-rebase-and-reformat-git-branches-automatically.md [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+:::info
+
+The reverts must have a fixup to target, lest they revert the real commit
+We could use `--allow-empty` for the fixup,
+this will guard against reverting a real commit.
+But git-revert will not create an empty commit.
+This is handled in `reformat.sh`
+
+:::
+
```
Comment (moroten) [inline blog/2023-12-10-rebase-and-reformat-git-branches-automatically.md:145]: Not needed any more.

File: blog/2023-12-10-rebase-and-reformat-git-branches-automatically.md [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+# r, reword <commit> = use commit, but edit the commit message
+# e, edit <commit> = use commit, but stop for amending
+# s, squash <commit> = use commit, but meld into previous commit
+# f, fixup [-C | -c] <commit> = like "squash" but keep only the previous
+#                    commit's log message, unless -C is used, in which case
+#                    keep only this commit's message; -c is same as -C but
+#                    opens the editor
+# x, exec <command> = run command (the rest of the line) using shell
+# ...
+```
```
Comment (moroten) [inline blog/2023-12-10-rebase-and-reformat-git-branches-automatically.md:76]: Move this to the second wave formatting.

File: blog/2023-12-10-rebase-and-reformat-git-branches-automatically.md [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+    cc2 Revert "fixup! fix/#123: real bazz"
+
+So `aaa` and `aa1` are combined into a single well-formatted commit `Aaa`,
+but the formatting is removed in the following `aa2`,
+so the original `bbb` still applies (but it is folded into `Bbb`).
+
+### Why not `--autosquash`?
+
+This is a good question, the `--autosquash` flag has long been a staple for me.
+But it does not work if there are duplicated commit messages in the commit chain.
```
Comment (moroten) [inline blog/2023-12-10-rebase-and-reformat-git-branches-automatically.md:178]:
```suggestion
This is a good question, the `--autosquash` flag has long been a staple for me,
but it does not work if there are duplicated commit messages in the commit chain.
```

File: blog/2023-12-10-rebase-and-reformat-git-branches-automatically.md [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+A general problem with such messages
+not specific to this workflow.
```
Comment (moroten) [inline blog/2023-12-10-rebase-and-reformat-git-branches-automatically.md:180]: Incomplete sentence, missing a verb.

File: blog/2023-12-10-rebase-and-reformat-git-branches-automatically.md [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+
+A less-than-imaginative "tidy up" commit message may trip up this flow.
```
Comment (moroten) [inline blog/2023-12-10-rebase-and-reformat-git-branches-automatically.md:182]: Why does "tidy up" trip up the flow?

File: blog/2023-12-10-rebase-and-reformat-git-branches-automatically.md [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+To make sure that all commits are up to the latest standard
+you can run the oneliner(s) from the end of this document.
+
+## Example commits
+
+Here are three example commits:
+
+    aaa feature 1: bazz the foo
+    bbb feature 2: bar bar
+    ccc fix/#123: real bazz
```
[the remark this answers is not in the document]
Comment (stagnation) [reply blog/2023-12-10-rebase-and-reformat-git-branches-automatically.md:25]: Good Idea! It makes tokenization easier for the visual system, and reduces repetitive patterns.

File: blog/2023-12-14-rebase-and-reformat-git-branches-automatically.md [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+  This reverts commit: $shaid"
+
+      git revert --no-commit HEAD
+      git commit --allow-empty -m "$revert_message"
+  }
+
+  tools/run.py 7..13
+  git add -u
+  git commit --allow-empty --fixup HEAD
+  revert
```
Comment (moroten) [inline blog/2023-12-14-rebase-and-reformat-git-branches-automatically.md:76]: It's better if we can keep all the additions to the `reformat.sh` script at the end.

```suggestion
  ./run-all-linters-and-auto-fixers
  
  # Add a new commit with the changes and revert it again.
  git add -u
  git commit --allow-empty --fixup HEAD
  # 'git-revert' does not support '--allow-empty'
  git revert --no-commit HEAD
  git commit --allow-empty --no-edit
```

File: docs/practice/rebase-and-reformat-git-branches-automatically.md [hunk 8]
```diff
@@ -0,0 +1,625 @@
...
+}
+
+# Formatters and fixers go here.
+# Replace with your tools of choice! rustfmt, gofmt, black, ...
+tools/run.py 7..13
+
+# Create more commits that we will squash later.
+git add -u
+git commit --allow-empty --fixup HEAD
+revert
```
Comment (moroten) [inline docs/practice/rebase-and-reformat-git-branches-automatically.md:80]: See the comment in the blog post.

File: blog/2023-12-14-rebase-and-reformat-git-branches-automatically.md [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+    EDITOR="sed -i '1,9d'"                                                                          \
+    GIT_SEQUENCE_EDITOR="vim +'g/^#/d' +'normal! Gdk' +'g/^pick \w* Revert \"fixup!/normal! j0ces'" \
+    git rebase -i origin/main
+```
+
+:::info
+We have not developed the *incantation*, `git-rebase` command,
+to preserve the author date from the original commits.
+We will address that next!
+:::
```
Comment (moroten) [inline blog/2023-12-14-rebase-and-reformat-git-branches-automatically.md:201]: I haven't verified the code. Unless you would like to try it, I'll try it another day.

```suggestion
:::

## Single pass update

Using a multi step approach is a good way to review the intermediate results. If you are feeling lucky, you may do it all in one pass, with correctness regards to keeping the original author information.

  ```bash title=reformat.sh
  #!/bin/sh

  ./run-all-linters-and-auto-fixers

  # Add a new commit with the changes and revert it again.
  git add -u
  git commit --allow-empty --fixup HEAD
  fixup_commit=$(git rev-parse HEAD)
  git reset --soft HEAD~
  original_commit=$(git rev-parse HEAD)
  if [ "$(git show -q --format=%s HEAD~)" = "revert-identifier" ]; then
    git reset --soft HEAD~
  fi
  git commit --amend --no-edit -c "$original_commit"

  # 'git-revert' does not support '--allow-empty'
  git revert --no-commit "$fixup_commit"
  git commit --allow-empty -m "revert-identifier"
  ` ` `

After the rebase, don't forget `git checkout HEAD~` to remove the last revert.
```

**Review by moroten — COMMENTED** [review]
Good flow. I'm happy with the structure on both blog and `docs/practice`.

**Comment by moroten** [comment]
Another variant is to run
```
git rebase -i origin/main --exec 'if [ "$(git show -q --format=%s)" = "fixup-identifier" ]; then git reset --soft HEAD~ && git commit --amend --no-edit ; elif [ "$(git show -q --format=%s HEAD~)" = "revert-identifier" ]; then head=$(git rev-parse HEAD) && git reset --soft HEAD~ && git commit --amend --no-edit -c $head; fi'
```

File: blog/2023-12-14-rebase-and-reformat-git-branches-automatically.md (hunk 7 above)
Comment (stagnation) [reply blog/2023-12-14-rebase-and-reformat-git-branches-automatically.md:76]: Ack, it _does_ create the commit message. Good that we do not have to.

File: docs/practice/rebase-and-reformat-git-branches-automatically.md [hunk 10]
```diff
@@ -0,0 +1,632 @@
...
+
+This opens the interactive rebase todo list with the `execute` commands filled in:
+
+```bash title=.git/rebase-merge/git-rebase-todo
+pick aaa My amazing feature
+x ./reformat.sh
+pick bbb Other complimentary work
+x ./reformat.sh
+pick ccc Fix annoying bug
+x ./reformat.sh
```
Comment (moroten) [inline docs/practice/rebase-and-reformat-git-branches-automatically.md:90]: In my editor it says `exec` and not just `x`.

File: docs/practice/rebase-and-reformat-git-branches-automatically.md [hunk 11]
```diff
@@ -0,0 +1,632 @@
...
+
+```
+(vim) /^pick \w* Revert "fixup!
+      ^                         : search
+       ^                        : Regexp: <beginning of line>pick <word> Revert "fixup!
+```
+
+This just searches for the revert-fixup commits,
+which should be roughly every other commit.
+There may be some gaps,
```
Comment (moroten) [inline docs/practice/rebase-and-reformat-git-branches-automatically.md:327]: No gaps any more. Actually, what happens if the fixup cancels the original commit? Then there is nothing for the revert to be squashed on to.

**Review by moroten — APPROVED** [review]
Optional to fix the comments. Let's do that in a later commit as it's not part of the blog.

File: docs/practice/rebase-and-reformat-git-branches-automatically.md (hunk 10 above)
Comment (stagnation) [reply docs/practice/rebase-and-reformat-git-branches-automatically.md:90]: Good point, I just use the shorthands, it is better to use the long form commands :)

**Pull request merged**